_posts/2022-09-07-storage-write-removal-before-conditional-termination.md

```diff
@@ -0,0 +1,161 @@
+2. A call to a function that conditionally terminates using inline assembly as described above, but may also regularly return to the caller.
+3. Either of the following:
+   a) Another (unconditional) storage write overwriting the write in (1).
+   b) An unconditional revert.
+
+With legacy code generation, all of this, i.e. steps 1-3, need to happen in a single inline assembly block and the function call in (2) needs to be a call to a user-defined assembly function (see below for examples of affected Yul snippets). Also note that during legacy code generation, the Yul optimizer is only run on inline assembly snippets that do not refer to Solidity variables, further reducing the amount of potentially affected cases.
+
+However, when using via-IR code generation the entire contract is first translated to Yul and then optimized as a whole. In this case, 1-3 can happen in solidity code and only the function called in (2) needs to involve an inline assembly block calling ``return(...)`` or ``stop()``. Note that due to inlining, this inline assembly block can also merely occur in another nested function call, i.e. any call in (2) that can both return to the caller and terminate via assembly is susceptible.
+
+The following is a minimal example that will exhibit the bug if compiled via-IR with enabled optimizer:
+
+```
+contract C {
+	uint public x;
+	function f(bool a) public {
+		x = 1; // This write used to be removed by the Yul optimizer.
+		g(a);
+		x = 2;
+	}
+	function g(bool a) internal {
+		// The relevant part of this function is that it can both return to the caller and terminate the transaction.
+		// The bug will show its effects in the cases in which the transaction is terminated (i.e. if a is false).
+		// In this case the write x = 1 above will be missing.
+		if (a) return;
+        assembly { return(0,0) }
+	}
+}
+```
+
+Compiling the above via-IR with enabled optimizer will result in ``f(false)`` to incorrectly terminate the transaction without modifying ``x``.
+
+For a concrete contract, the bug may still be prevented, in case the optimizer inlines the function before running the problematic optimizer step, e.g. the following contract is almost identical, but unaffected, since ``g()`` will be inlined (in the snippet above the ``return;`` prevents easy inlining):
+
+```
+contract C {
+	uint public x;
+	function f(bool a) public {
+		x = 1;
+		g(a);
+		x = 2;
+	}
+	function g(bool a) internal {
+		if (!a)
+			assembly { return(0,0) }
+	}
+}
+```
+
+To check whether your contract is actually affected, you need to trigger the condition leading to the assembly ``return(...)`` or ``stop()`` and verify that all storage writes that should have happened in the code leading to it are correctly performed.
+
+   
+## Technical Details
+
+The Yul optimizer step responsible for the bug is meant to remove storage writes that it can determine to be redundant. A storage write is considered redundant in either of the following two cases:
```

Comment:
Optimizer steps are documented in the docs so it would probably be a good idea to mention its name and abbreviation. 

Also maybe say that one way to completely mitigate this bug with earlier compilers would be to set a custom sequence without this step?

Reply:
Do we want to mention that a custom sequence potentially mitigates it?